api/konnect/v1alpha1/konnect_cloudgateway_dataplane_configuration_types.go

```diff
@@ -0,0 +1,257 @@
+	// ConfigurationDataPlaneGroupAutoscaleStatic specifies the static configuration for the data-plane group.
+	//
+	// +kubebuilder:validation:Optional
+	ConfigurationDataPlaneGroupAutoscaleStatic *ConfigurationDataPlaneGroupAutoscaleStatic `json:"static,omitempty"`
+
+	// ConfigurationDataPlaneGroupAutoscaleAutopilot specifies the autoscale configuration for the data-plane group.
+	//
+	// +kubebuilder:validation:Optional
+	ConfigurationDataPlaneGroupAutoscaleAutopilot *ConfigurationDataPlaneGroupAutoscaleAutopilot `json:"autopilot,omitempty"`
+
+	// Type of autoscaling to use.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=ConfigurationDataPlaneGroupAutoscaleStatic;ConfigurationDataPlaneGroupAutoscaleAutopilot
+	Type sdkkonnectcomp.ConfigurationDataPlaneGroupAutoscaleType `json:"type"`
+}
+
+// ConfigurationDataPlaneGroupAutoscaleAutopilot specifies the autoscale configuration for the data-plane group.
+type ConfigurationDataPlaneGroupAutoscaleAutopilot struct {
+	// Base number of requests per second that the deployment target should support.
+	//
+	// +kubebuilder:validation:Required
+	BaseRps int64 `json:"base_rps"`
+
+	// Max number of requests per second that the deployment target should support. If not set, this defaults to 10x base_rps.
+	MaxRps *int64 `json:"max_rps,omitempty"`
+}
+
+// ConfigurationDataPlaneGroupAutoscaleStatic specifies the static configuration for the data-plane group.
+type ConfigurationDataPlaneGroupAutoscaleStatic struct {
+	// Instance type name to indicate capacity.
+	//
+	// +kubebuilder:validation:Required
+	InstanceType sdkkonnectcomp.InstanceTypeName `json:"instance_type"`
+
+	// Number of data-planes the deployment target will contain.
+	//
+	// +kubebuilder:validation:Required
+	RequestedInstances int64 `json:"requested_instances"`
+}
+
+// KonnectCloudGatewayDataPlaneGroupConfigurationStatus defines the observed state of KonnectCloudGatewayDataPlaneGroupConfiguration.
+// +apireference:kgo:include
+type KonnectCloudGatewayDataPlaneGroupConfigurationStatus struct {
+	KonnectEntityStatusWithControlPlaneRef `json:",inline"`
+
+	// DataPlaneGroups is a list of deployed data-plane groups.
+	DataPlaneGroups []KonnectCloudGatewayDataPlaneGroupConfigurationStatusGroup `json:"dataplane_groups,omitempty"`
+
+	// Conditions describe the current conditions of the KonnectCloudGatewayDataPlaneGroupConfiguration.
+	//
+	// Known condition types are:
+	//
+	// * "Programmed"
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// KonnectCloudGatewayDataPlaneGroupConfigurationStatusGroup defines the observed state of a deployed data-plane group.
+type KonnectCloudGatewayDataPlaneGroupConfigurationStatusGroup struct {
```

Comment:
In case this field is intended to be created and set in the same controller's iteration, I would recommend marking the fields below as required. I don't see much value in having them optional. WDYT?

Reply:
I'm not sure if filling the status would all happen in the same reconciliation iteration (thinking about it, it does make sense that it would) and marking it as such could preclude e.g patching with incomplete information.


>  I don't see much value in having them optional

I do see your concern about this but at the same time: "I don't see much value in having them required". Any particular reason to make this required?

Reply:
👍 9eaf2e981f21713caec0652eae0bf4e630ed0899